scripts/run_tests.sh

```diff
@@ -0,0 +1,131 @@
+if ( nc -zv $IP_ADDRESS $JUPYTER_PORT 2>&1 >/dev/null ); then
+    echo "The address ${IP_ADDRESS}:${JUPYTER_PORT} is already in use; please kill the existing notebook server to run tests."
+    exit 127
+fi
+
+if [ "$run_docker" == 1 ]; then
+    echo "starting narrative container ${CONTAINER}"
+    rm -f $OUTPUT_FILE
+    touch $OUTPUT_FILE
+    docker run -p $JUPYTER_PORT:$JUPYTER_PORT $CONTAINER \
+        --template \
+        "/kb/dev_container/narrative/src/config.json.templ:/kb/dev_container/narrative/src/config.json" \
+        --template \
+        "/kb/dev_container/narrative/src/config.json.templ:/kb/dev_container/narrative/kbase-extension/static/kbase/config/config.json" \
+        kbase-narrative --no-browser --NotebookApp.allow_origin="*" --ip=$IP_ADDRESS --port=$JUPYTER_PORT 2>&1 | tee $OUTPUT_FILE &
+    bg_pid=$!
+else
+    echo "starting local narrative"
+    rm -f $OUTPUT_FILE
+    touch $OUTPUT_FILE
+    kbase-narrative --no-browser --NotebookApp.allow_origin="*" --ip=$IP_ADDRESS --port=$JUPYTER_PORT 2>&1 | tee $OUTPUT_FILE &
+    bg_pid=$!
+fi
+
+while :; do
+    if grep -q "is already in use" $OUTPUT_FILE; then
+        echo "The port ${JUPYTER_PORT} is already in use; please kill the existing notebook server to run tests."
+        kill $bg_pid
+        exit 127
+    fi
+    if grep -q "port is already allocated" $OUTPUT_FILE; then
+        echo "The port ${JUPYTER_PORT} is already in use; please kill the existing notebook server to run tests."
+        kill $bg_pid
+        exit 127
+    fi
+    if grep -q "The Jupyter Notebook is running at:" $OUTPUT_FILE; then
+        # exit the loop once the string has been found
+        echo "narrative started"
+        break;
+    fi
+    sleep 1
+done
+
+echo "starting tests"
+
+isolated=-1
+unit=-1
+integration=-1
+
+if [ "${run_unit}" == 1 ]; then
+    echo "running isolated unit tests"
+    npm run test_isolated
+    isolated=$?
+
+    echo "running main unit tests"
+    npm run test
+    unit=$?
+fi
+
+if [ "${run_integration}" == 1 ]; then
+    base_url=${BASE_URL:-"http://localhost:$JUPYTER_PORT"}
+    export BASE_URL=$BASE_URL
+    npx wdio test/integration/wdio.conf.js
+    integration=$?
+fi
+
+echo "Finished running tests"
+
+declare -a names=("isolated" "unit" "integration")
+declare -a results=($isolated $unit $integration)
+length=${#results[@]}
+
+for (( i=0; i < ${length}; i++ ))
+do
+    if [ "${results[$i]}" != -1 ]; then
+        echo "${names[$i]} tests completed with code ${results[$i]}"
+        exit_code=$(( exit_code + results[$i] ))
+    fi
+done
+
+echo "Killing server"
+
+kill_descendant_processes() {
+    local pid="$1"
+    local and_self="${2:-false}"
+    if children="$(pgrep -P "$pid")"; then
+        for child in $children; do
+            kill_descendant_processes "$child" true
+        done
+    fi
+    if [[ "$and_self" == true ]]; then
+        kill -9 "$pid"
+    fi
+}
+
+kill_descendant_processes $bg_pid
```

Comment:
This doesn't work properly -- the docker container doesn't get shut down when the script exits. For the purposes of GHA, the tests are set up such that it doesn't matter if a container is left running. If there are any shell scripters who want to fix this, they are more than welcome to do so!